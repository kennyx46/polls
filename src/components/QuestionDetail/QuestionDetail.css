.activeRow {
    background-color: rgba(0, 0, 0, 0.075);
}

.progressBarItem {
	min-width: 80px;
}
