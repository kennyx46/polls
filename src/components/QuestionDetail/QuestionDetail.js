import React, { Component } from 'react';
import Container from 'react-bootstrap/Container';
import Table from 'react-bootstrap/Table';
import ProgressBar from 'react-bootstrap/ProgressBar';
import Button from 'react-bootstrap/Button';
import { Link } from 'react-router-dom';

import './QuestionDetail.css';

export default class QuestionDetail extends Component {

    state = {
        selectedChoiceIndex: -1,
    }

    componentDidMount() {
        const { questionId } = this.props.match.params;
         const { question } = this.props;
        if (!question) {
             this.props.getQuestion({ questionId });
        }
    }

    renderChoice = (choice, index) => {
        const { selectedChoiceIndex } = this.state;

        return (
            <tr key={choice.url} 
                className={index === selectedChoiceIndex ? 'activeRow' : ''} 
                onClick={() => this.setState({selectedChoiceIndex: index})}>
                <td>{choice.choice}</td>
                <td>{choice.votes}</td>
                <td>{choice.percentage}</td>
                <td className="progressBarItem">
                    <ProgressBar now={choice.percentage}/>
                </td>
            </tr>
        )
    }

    saveVote = () => {
        const { choices } = this.props;
        const { selectedChoiceIndex } = this.state;
        if (selectedChoiceIndex > -1) {
            this.props.voteOnChoice({ choiceUrl: choices[selectedChoiceIndex].url });
        }
    }

    render() {
        const { isLoading, question, choices, isVotingInProgress } = this.props;

        if (isLoading || !question) {
            return (
                <Container>
                    <p>Loading</p>
                </Container>
            );
        }

        return (
            <Container>
                <h1>Questions Detail</h1>
                <div className="my-2">
                    <Link to='/'>Questions</Link>
                </div>
                <h2>Question: {question.question}</h2>
                <Table bordered hover>
                    <tbody>
                        {choices.map(this.renderChoice)}
                    </tbody>
                </Table>
                <Button disabled={isVotingInProgress} className="float-right" onClick={this.saveVote}>Save vote</Button>
            </Container>
        );
    }
}
