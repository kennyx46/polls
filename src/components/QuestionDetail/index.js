import { connect } from 'react-redux';
import { compose } from 'redux';
import { getQuestion, voteOnChoice, } from '../../data/questions/actions';
import QuestionDetail from './QuestionDetail';
import { getCurrentQuestionChoicesWithPercentage } from '../../data/questions/selectors';
import { withRouter } from "react-router";

const mapStateToProps = (state) => ({
    question: state.questions.current,
    choices: getCurrentQuestionChoicesWithPercentage(state),
    isLoading: state.questions.isLoading,
    isVotingInProgress: state.questions.isVotingInProgress,
});

const mapDispatchToProps = {
    getQuestion,
    voteOnChoice,
};

export default compose(
	connect(mapStateToProps, mapDispatchToProps),
	withRouter,
)(QuestionDetail);
