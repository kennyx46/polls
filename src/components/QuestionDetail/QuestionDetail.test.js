import React from 'react';
import { shallow } from 'enzyme';
import Card from 'react-bootstrap/Card';
import Button from 'react-bootstrap/Button';

import QuestionDetail from './QuestionDetail';

describe('QuestionDetail', () => {
    const defaultProps = Object.freeze({
        question: {},
        choices: [],
        isLoading: false,
        isVotingInProgress: false,
        error: null,
        match: { params: {} },
        voteOnChoice: jest.fn(),
    });

    it('renders component', async () => {
        const wrapper = shallow(
            <QuestionDetail {...defaultProps }/>
        ).dive();
        expect(wrapper).toMatchSnapshot();
    });

    it('shows loading if loading in progress', async () => {
        const props = { ...defaultProps, isLoading: true };
        const questionDetail = shallow(<QuestionDetail {...props} />);
        expect(questionDetail.text()).toContain('Loading');
    });

    it('renders choices for a question', async () => {
        const sampleChoice = { url: '1' };
        const props = { ...defaultProps, choices: [sampleChoice, { ...sampleChoice, url: '2' }, { ...sampleChoice, url: '3' }] };
        const questionDetail = shallow(<QuestionDetail {...props} />);
        expect(questionDetail.find('tr')).toHaveLength(3);
    });

    it('saves vote', async () => {
        const sampleChoice = { url: '1' };
        const props = { ...defaultProps, choices: [sampleChoice, { ...sampleChoice, url: '2' }, { ...sampleChoice, url: '3' }] };
        const questionDetail = shallow(<QuestionDetail {...props} />);
        questionDetail.find('tr').at(1).simulate('click');
        questionDetail.find(Button).simulate('click');
        expect(defaultProps.voteOnChoice).toHaveBeenCalled();
    });

});