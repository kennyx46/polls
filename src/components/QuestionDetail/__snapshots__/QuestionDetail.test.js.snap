// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuestionDetail renders component 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <ForwardRef(Bootstrap(Container))>
    <h1>
      Questions Detail
    </h1>
    <div
      className="my-2"
    >
      <Link
        replace={false}
        to="/"
      >
        Questions
      </Link>
    </div>
    <h2>
      Question: 
    </h2>
    <ForwardRef(Bootstrap(Table))
      bordered={true}
      hover={true}
    >
      <tbody />
    </ForwardRef(Bootstrap(Table))>
    <ForwardRef(Bootstrap(Button))
      className="float-right"
      disabled={false}
      onClick={[Function]}
    >
      Save vote
    </ForwardRef(Bootstrap(Button))>
  </ForwardRef(Bootstrap(Container))>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": [Function],
    },
    "ref": null,
    "rendered": [Function],
    "type": Object {
      "$$typeof": Symbol(react.context),
      "_calculateChangedBits": null,
      "_context": Object {
        "$$typeof": Symbol(react.context),
        "Consumer": [Circular],
        "Provider": Object {
          "$$typeof": Symbol(react.provider),
          "_context": [Circular],
        },
        "_calculateChangedBits": null,
        "_currentRenderer": null,
        "_currentRenderer2": null,
        "_currentValue": Map {},
        "_currentValue2": Map {},
        "_threadCount": 0,
      },
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": [Function],
      },
      "ref": null,
      "rendered": [Function],
      "type": Object {
        "$$typeof": Symbol(react.context),
        "_calculateChangedBits": null,
        "_context": Object {
          "$$typeof": Symbol(react.context),
          "Consumer": [Circular],
          "Provider": Object {
            "$$typeof": Symbol(react.provider),
            "_context": [Circular],
          },
          "_calculateChangedBits": null,
          "_currentRenderer": null,
          "_currentRenderer2": null,
          "_currentValue": Map {},
          "_currentValue2": Map {},
          "_threadCount": 0,
        },
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "context": Object {},
  },
}
`;
